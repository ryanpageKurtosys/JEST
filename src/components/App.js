import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import Gift from './Gift';
import { max_number } from '../helper';

class App extends Component {

    constructor(){
        super();

        this.state = { 
            gifts: [],
            width: 0, 
            height: 0,

        };
    }

    addGift = () => {

        const { gifts } = this.state;

        const ids = this.state.gifts.map((gift) => {
           return gift.id
        });

        gifts.push({ id: max_number(ids)+1 });

        this.setState({gifts});

    }

    removeGift = id => {
        const gifts = this.state.gifts.filter(gift => gift.id !== id);

        this.setState({ gifts });
    }

    render(){
        return(
            <div>
                 <nav>
                    <div className ="nav-wrapper cyan darken-3 z-depth-1">
                        <a href='#!' style={{paddingLeft:20}} className="brand-logo">Gift Giver</a>
                    </div>
                </nav>

                <div style={{justifyContent:'center', paddingLeft:'25%', paddingRight:'25%' }}>
                <h2 style={{display: 'flex', justifyContent: 'center'}}>Gift Giver</h2>
                <div className='gift-list' style={{paddingBottom:20}}>
                    {
                    this.state.gifts.map(gift => {
                            return (
                                <Gift 
                                    key={gift.id}
                                    gift={gift}
                                    removeGift={this.removeGift}
                                    >
                                </Gift>
                            )
                        })
                    }
                    </div>
                    <div class='section' style={{display: 'flex', justifyContent: 'center', paddingTop:30}}>
                        <Button  className='btn-add' onClick={this.addGift}>Add Gift</Button>
                    </div>
                    
                </div>
            </div>
        )
    }
}

export default App; 