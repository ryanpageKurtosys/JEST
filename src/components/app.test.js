import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import renderer from 'react-test-renderer';


describe('App', () => {
    //helper function shallowly renders the react component ( only renders the parent )
    const app = shallow(<App />);

    it('renders correctly', () => {
        //enzyme method
        var wrapper = renderer.create(<App />).toJSON();
        //toMatchSnapshot -jest feature. keep a record history of react components ( can tell when a components render output changes)
        // this is done so that if changes are made they you can see if they have an effect on the rendered component
        expect(wrapper).toMatchSnapshot();
    });
    
    it('initialize the state with an emply list of gifts' , () => {
        //expect checks that parts of the react component equal a certain criteria
        //app.state() - this allows you get get the state of the react component
        // this is checking that when the component renders the initail state is an empty array
        expect(app.state().gifts).toEqual([]);
        //expect(app.state().gifts).toEqual([{ id:3 }]);
    });
    
    //Open a describe because all the following tests pretain to the same thing ( the button )
    describe('when clicking the `add gift` button', () => {

        const id = 1;
        // This sets up a method that is called at every `it` instance this code runs before the code in the `it`        
        beforeEach(() => {
            //app.find is used to get hold of inner child nodes/className from the react component using jsx tags
            //enzyme method to simulate a click
            app.find('.btn-add').simulate('click');
            //gift.find('.input-present').simulate('change', { target: { value: present}});

        });

        // This sets up a method that is called at every `it` instance this code runs after the code in the `it`        
        afterEach(() => {
            app.setState({ gifts:[] });
        });

        //example without the before each and after each 
        // it(`adds a new gift to 'state'`,  () => {
        //     app.find('.btn-add').simulate('click');
        //     expect(app.state().gifts).toEqual([{ id }]);
        //     app.setState({ gifts:[] });
        // });

        it(`adds a new gift to 'state'`,  () => {
            // this will check the state after the button is clicked 
            expect(app.state().gifts).toEqual([{ id }]);
        });

        it('creates a gift component', () => {
            //checking that the Gift react component renders
            expect(app.find('Gift').exists()).toBe(true);
        });

        describe('and the user wants to remove the added gift', () => {
            beforeEach(() => {
                //this selects the remove gift button
                app.instance().removeGift(id);
            });

            it('removes gift from state',() => {
                // this checks to see after the button is clicked is the gift removed
                expect(app.state().gifts).toEqual([]);
            });
        });
    
    }) 
})

