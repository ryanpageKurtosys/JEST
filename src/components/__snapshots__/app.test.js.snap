// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <nav>
    <div
      className="nav-wrapper cyan darken-3 z-depth-1"
    >
      <a
        className="brand-logo"
        href="#!"
        style={
          Object {
            "paddingLeft": 20,
          }
        }
      >
        Gift Giver
      </a>
    </div>
  </nav>
  <div
    style={
      Object {
        "justifyContent": "center",
        "paddingLeft": "25%",
        "paddingRight": "25%",
      }
    }
  >
    <h2
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      Gift Giver
    </h2>
    <div
      className="gift-list"
      style={
        Object {
          "paddingBottom": 20,
        }
      }
    />
    <div
      class="section"
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "paddingTop": 30,
        }
      }
    >
      <button
        className="btn-add btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Add Gift
      </button>
    </div>
  </div>
</div>
`;
